tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,128 @@
+#ifndef ROOT_RDF_RMERGEABLEVALUE
+#define ROOT_RDF_RMERGEABLEVALUE
+
+#include <functional>
+#include <memory>
+
+#include "RtypesCore.h"
+#include "TList.h"
+
+namespace ROOT {
+namespace RDF {
+
+template <typename T>
+class RMergeableValue {
+protected:
+   T fValue;
+
+public:
+   RMergeableValue(const T &value) : fValue{value} {}
+   virtual ~RMergeableValue() = default;
+   virtual void Merge(const std::unique_ptr<RMergeableValue<T>> &) = 0;
+   virtual const T &GetValue() const { return fValue; }
+};
+
+// Specialization for generic action
+template <typename T>
+class RMergeableGenericAction final : public RMergeableValue<T> {
+   using mergefun_t = std::function<T(const T &, const T &)>;
+   const mergefun_t fMergeFun;
+
+public:
+   RMergeableGenericAction(const T &value, const mergefun_t &mergefun) : RMergeableValue<T>(value), fMergeFun{mergefun}
+   {
+   }
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      const auto &othervalue = other->GetValue();
+      this->fValue = fMergeFun(this->fValue, othervalue);
+   }
+};
+
+// RMergeableValue specialization for the Mean action
+template <typename T>
+class RMergeableMean final : public RMergeableValue<T> {
+   ULong64_t fCounts;
+
+public:
+   RMergeableMean(const T &value, const ULong64_t &counts) : RMergeableValue<T>(value), fCounts{counts} {}
+
+   const ULong64_t &GetCounts() const { return fCounts; }
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      const auto &othervalue = other->GetValue();
+      const auto &othercounts = dynamic_cast<const RMergeableMean<T> &>(*other).GetCounts();
+      const auto num = this->fValue * fCounts + othervalue * othercounts;
+      const auto denum = static_cast<double>(fCounts + othercounts);
+      this->fValue = num / denum;
+      fCounts += othercounts;
+   }
+};
+
+template <typename T>
+class RMergeableHisto1D final : public RMergeableValue<T> {
+public:
+   RMergeableHisto1D(const T &value) : RMergeableValue<T>(value) {}
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final { this->fValue.Add(&(other->GetValue())); }
+};
+
+template <typename T>
+class RMergeableFillParHelper final : public RMergeableValue<T> {
+public:
+   RMergeableFillParHelper(const T &value) : RMergeableValue<T>(value) {}
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      TList l;
+      l.Add(const_cast<T *>(&(other->GetValue())));
```

Comment:
ugh...what type does not take a `const T *` as argument?

Reply:
aaah I misread, it's `TList`...well, I guess we can't do much about this.

Reply:
`void TList::Add(TObject *obj) { AddLast(obj); }` :sweat_smile:

Reply:
See also https://github.com/root-project/root/blob/768d59827c9ecaca7f34c5c9c5348305411d14cb/hist/hist/src/TH1.cxx#L901-L906

Reply:
yes you are right, see my second comment in this thread, conversation was already resolved :+1: